lib/dataformat/dbmon/dbmon.3
.TH DBMON 3
.SH NAME
dbmon_init, dbmon_resync, dbmon_update, dbmon_status, dbmon_close, dbmon_compute_row_sync \- monitor a Datascope database for changes
.SH SYNOPSIS
.nf

#include "dbmon.h"

Hook *\fBdbmon_init\fP( Dbptr \fIdb\fP, Tbl *\fItable_subset\fP, 
                  void (*\fInewrow\fP)(Dbptr db, char *table, long irecord, char *sync, void *pvt), 
                  void (*\fIdelrow\fP)(Dbptr db, char *table, char *oldsync, void *pvt), 
                  Tbl *(*\fIquerysyncs\fP)(Dbptr db, char *table, void *pvt),
                  int \fIflags\fP );

int \fBdbmon_resync\fP( Hook *\fIdbmon_hook\fP, void *\fIpvt\fP );

int \fBdbmon_update\fP( Hook *\fIdbmon_hook\fP, void *\fIpvt\fP );

void \fBdbmon_status\fP( FILE *\fIfp\fP, Hook *\fIdbmon_hook\fP );

void \fBdbmon_close\fP( Hook **\fIdbmon_hook\fP );

char *\fBdbmon_compute_row_sync\fP( Dbptr \fIdb\fP );

.fi
.SH DESCRIPTION
The \fBdbmon_init\fP function begins tracking a dynamically updating
Datascope database. The \fIdb\fP database pointer indicates the
database to be tracked. If the \fItable_subset\fP is null, all
tables are tracked.  If this argument is a list of string names for
database tables (see strtbl(3)), tracking is limited to the named
tables.  \fBdbmon_init\fP returns a hook(3) pointer which is used
in subsequent calls to interact with the tracked monitoring status
of the database. Three callbacks are registered when invoking
\fBdbmon_init\fP: \fInewrow\fP, \fIdelrow\fP, and \fIquerysyncs\fP.
The pointer to the \fIquerysyncs\fP callback function is allowed to be 
\fINULL\fP, effectively disabling the \fBdbmon_resync\fP function described
below. 

After calling \fBdbmon_init\fP, the calling program may call the \fBdbmon_resync\fP
function, if desired, to compare results from past monitoring with 
the current database and respond to any changes.  After that, the calling 
program must call the \fBdbmon_update\fP function at desired intervals to trigger 
updates based on changes to the monitored database. (If the calling 
program omits the call to \fBdbmon_resync\fP, the first call
to \fBdbmon_update\fP assumes the calling application knows nothing about 
the monitored database).

If the \fBdbmon_resync\fP function is called, the \fIquerysyncs\fP callback 
is invoked for each monitored table, asking the calling program for 
a table of the synchronization strings the program already knows about 
for that table (presumably from state saved by a previous execution of 
the program). The \fIquerysyncs\fP function is given the database pointer 
for the database (merely for informational purposes; in most cases said pointer
would not be used), the name of the table for which the sync strings are 
being requested, and the private pointer \fIpvt\fP which was passed 
to \fBdbmon_resync\fP. The \fIquerysyncs\fP callback must return a 
newly allocated \fITbl\fP (see tbl(3)) containing newly allocated copies of all known 
synchronization strings for the table in question After using said 
\fITbl\fP, \fBdbmon_resync\fP will free it with a call to 
.nf

freetbl( atbl, free );

.fi

For each Datascope database row newly detected by \fBdbmon_update\fP,
a database pointer pointing to the scratch record,
which will contain a copy of the new row, is given by \fBdbmon_update\fP as the first argument
to the \fInewrow\fP callback. The second argument to the \fInewrow\fP
callback is the Datascope table-name containing the row in question.
The third argument is the actual row number of the row in the table.
The fourth argument is a string computed by \fBdbmon_update\fP
intended to be a unique signature of the database row. This can be used
as desired by the calling application to identify the returned row
in future operations. The fifth argument to the \fInewrow\fP callback
is a void pointer \fIpvt\fP, the same as the \fIpvt\fP
pointer provided to the \fBdbmon_update\fP function. This private
pointer may be used by the application programmer writing callbacks
to handle more complex behavior for each database row, for example
the \fIpvt\fP pointer may reference a data structure for
transformed rows or instructions to another utility.

For database rows that \fBdbmon_update\fP concludes have been marked
for deletion, the \fIdelrow\fP callback is invoked, with the
\fIoldsync\fP string holding the value of the synchronization string
from before the row was marked. The \fItable\fP-name argument
indicates the database table from which the row was deleted. The
database pointer is provided as a convenience to the programmer and
does not contain any information about the row from before it was
marked for deletion.

The \fBdbmon_status\fP function is a debugging utility which prints
what is known internal to the library about the current status of
the database represented by the tracking hook.

The \fBdbmon_close\fP function stops tracking a previously tracked
Datascope database, freeing all tracking structures.

The \fBdbmon_compute_row_sync\fP function computes a synchronization
string for a given database row, used internally for tracking
functions and possibly useful externally as well to coordinate with
other operations. The returned character string should be freed by
the caller when no longer needed.

.SH EXAMPLE
The example below, taken from the certify(1) tests for the dbmon
library, shows a database being monitored, with some artificial
updates accomplished via system calls to copy alternate versions
of one of the database tables over the initial copy:
.in 2c
.ft CW
.nf

%cat certify/try_dbmon.c

#include <stdlib.h>
#include "db.h"
#include "stock.h"
#include "dbmon.h"
#include "db2sql.h"

void newrow( Dbptr db, char *table, long irecord, char *sync, void *pvt );
void delrow( Dbptr db, char *table, char *sync, void *pvt );

void
newrow( Dbptr db, char *table, long irecord, char *sync, void *pvt )
{ 
	char	row[10*STRSZ];
	FILE	*fp = (FILE *) pvt;

	dbget( db, row );
	
	fprintf( fp, "New Row %ld in '%s' [sync '%s']: %s\\n", irecord, table, sync, row );

	return;
}

void
delrow( Dbptr db, char *table, char *sync, void *pvt )
{ 
	FILE	*fp = (FILE *) pvt;

	fprintf( fp, "Deleted row from '%s' with sync '%s'\\n", table, sync );

	return;
}

int
main(int argc, char **argv )
{
	Dbptr	db;
	char	dbname[FILENAME_MAX];
	Hook	*dbmon_hook = NULL;
	Tbl	*tables = (Tbl *) NULL; 

	if( argc != 2 ) {

		elog_die( 0, "Usage: %s dbname\\n", argv[0] );

	} else {
		
		strcpy( dbname, argv[1] );
	}

	dbopen_database( dbname, "r", &db );

	dbmon_hook = dbmon_init( db, tables, newrow, delrow, NULL, 0 );

	dbmon_update( dbmon_hook, (void *) stdout );

	fprintf( stdout, "Adding new arrival table:\\n" );

	system( "cp data/mod.new.demo.arrival results/dbmon/demo.arrival" );

	sleep( 1 );

	dbmon_update( dbmon_hook, (void *) stdout );

	fprintf( stdout, "Shortening arrival table:\\n" );

	system( "cp data/mod.shorter.demo.arrival results/dbmon/demo.arrival" );

	sleep( 1 );

	dbmon_update( dbmon_hook, (void *) stdout );

	fprintf( stdout, "Lengthening arrival table:\\n" );

	system( "cp data/mod.longer.demo.arrival results/dbmon/demo.arrival" );

	sleep( 1 );

	dbmon_update( dbmon_hook, (void *) stdout );

	dbmon_status( stdout, dbmon_hook );

	clear_register( 1 );

	free_hook( &dbmon_hook );

	dbclose( db );

	exit( 0 );
}
.fi
.ft R
.in
.SH RETURN VALUES
\fBdbmon_init\fP returns a hook(3) pointer used in subsequent calls to the dbmon library. 

\fBdbmon_update\fP returns 0. \fBdbmon_status\fP and \fBdbmon_close\fP do not return anything. 

\fBdbmon_compute_row_sync\fP returns a char pointer with a string representation of the sync value 
for the given database row. The pointer must be freed by the caller. 
.SH LIBRARY
-ldbmon
.SH "SEE ALSO"
.nf
perldbmon(3p), db2sql(3)
.fi
.SH "BUGS AND CAVEATS"
The \fIflags\fP argument to \fBdbmon_init\fP is not yet used and should be set to 0. 

If the database descriptor-file changes such that there is a change
to which underlying files are referenced for each database table,
the dbmon library will not adjust dynamically.

For databases being modified and read by multiple processes (e.g.
real-time databases), rows for deletion should be marked with
dbmark(3) and left in place. The \fBdbmon_update\fP function does
assume the \fBorb2db_msg\fP program is on the execution path, running
that command with the \fIwait\fP option on the database before
proceeding with each table.  This allows other real-time processes
to execute cleanup operations provided they invoke \fBorb2db_msg(1)\fP
before doing so.

The dbmon library does not yet monitor external files, though perhaps should. 

In addition to the database pointer for a given row, the corresponding
table name is passed to the callback functions, even though the
table name could also be derived from a dbquery(3) call on the
database pointer. This is done as a convenience to the programmer
implementing the callbacks.

The \fInewrow\fP callback is given scratch records with copies of
the new rows, rather than being given the new rows themselves, in
order to keep the transactions atomic: i.e. to make sure the operation
performed on the database row in the callback is done on exactly
the same row for which the sync string was computed internal to the
dbmon library. This is necessary for databases that may be changing
more rapidly than the refresh rate of the dbmon library. An alternative
solution would be to force all other database-write operations to
suspend during a dbmon refresh operation, however that demands a
more stringent trade-off between real-time process latency and dbmon
latency, which seems less desirable.


.SH AUTHOR
.nf
Dr. Kent Lindquist 
Lindquist Consulting, Inc. 
.fi


lib/dataformat/dbmon/dbmon.c

/*
 *   Copyright (c) 2009-2011 Lindquist Consulting, Inc.
 *   All rights reserved. 
 *                                                                     
 *   Written by Dr. Kent Lindquist, Lindquist Consulting, Inc. 
 *
 *   This software is licensed under the New BSD license: 
 *
 *   Redistribution and use in source and binary forms,
 *   with or without modification, are permitted provided
 *   that the following conditions are met:
 *   
 *   * Redistributions of source code must retain the above
 *   copyright notice, this list of conditions and the
 *   following disclaimer.
 *   
 *   * Redistributions in binary form must reproduce the
 *   above copyright notice, this list of conditions and
 *   the following disclaimer in the documentation and/or
 *   other materials provided with the distribution.
 *   
 *   * Neither the name of Lindquist Consulting, Inc. nor
 *   the names of its contributors may be used to endorse
 *   or promote products derived from this software without
 *   specific prior written permission.

 *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 *   WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 *   PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 *   THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY
 *   DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 *   IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *   USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *   POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <stdlib.h>
#include "db.h"
#include "stock.h"
#include "crc.h"
#include "dbmon.h"

typedef struct Dbtrack {
	char	dbname[FILENAME_MAX];
	Dbptr	db;
	Arr	*tables;
	void 	(*newrow)(Dbptr, char *, long, char *, void *);
	void 	(*delrow)(Dbptr, char *, char *, void *);
} Dbtrack;

typedef struct Tabletrack {
	char	table_name[STRSZ];
	char	table_filename[FILENAME_MAX];
	Dbptr	db;
	int	table_exists;
	long	table_nrecs;
	unsigned long table_modtime;
	char	*null_sync;
	Stbl	*syncs;
	int	watch_table;
} Tabletrack;

typedef struct Synctrack {
	char	*sync;
	int	keep;
	int	add;
	int	irecord;
} Synctrack;

typedef struct SynctrackCtxt {
	Dbtrack *dbtr;
	Tabletrack *ttr;
	void	*pvt;
} SynctrackCtxt;

static Dbtrack *new_dbtrack( Dbptr db );
static Tabletrack *new_tabletrack( char *table_name );
static Synctrack *new_synctrack( Dbptr db, long irecord );
static SynctrackCtxt *new_synctrack_context( Dbtrack *dbtr, Tabletrack *ttr, void *pvt );
static int sort_byirecord( char **ap, char **bp, void *pvt );
static int applystbl_sorted( Stbl *stbl, int (*function)(void *, void *), void *pvt );
static int cmp_synctracks( void *ap, void *bp );
static int synctrack_reset_conditionals( void *strp, void *pvt );
static int synctrack_conditional_add( void *strp, void *strcxtp );
static int synctrack_conditional_delete( void *strp, void *strcxtp );
static int synctrack_certain_delete( void *strp, void *strcxtp );
static int synctrack_print( void *strp, void *fpp );
static int compute_digest( unsigned char *buf, unsigned int len, unsigned char *digest );
static char *digest2hex( unsigned char *digest );
static void free_tabletrack( void *ttrp );
static void free_dbtrack( void *dbtrp );
static void free_synctrack( void *strp );
static void free_synctrack_context( SynctrackCtxt *strp );
static void focus_tableset( Dbtrack *dbtr, Tbl *table_subset );
static void dbmon_build_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt );
static void dbmon_update_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt );
static void dbmon_delete_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt );

static int 
compute_digest( unsigned char *buf, unsigned int len, unsigned char *digest )
{
	struct sha_ctx ctx;

	sha_init( &ctx );
	sha_update( &ctx, (unsigned char *) buf, len );
	sha_final( &ctx );
	sha_digest( &ctx, digest );

	return 0;
}

static char *
digest2hex( unsigned char *digest )
{
	char	*hex = NULL;
	int	i = 0;

	allot( char *, hex, 41 );
	
	for( i=0; i<20; i++ ) {
		
		sprintf( &hex[2*i], "%02x", digest[i] );
	}

	hex[40] = '\0';

	return hex;
}

static Dbtrack *
new_dbtrack( Dbptr db )
{
	Dbtrack *dbtr = NULL;
	Dbvalue val;
	Tabletrack *ttr = NULL;
	Tbl	*schema_tables = NULL;
	char	*table_name = NULL;
	long	itable = 0L;

	allot( Dbtrack *, dbtr, 1 );

	dbtr->db = db;
	dbtr->db.table = dbALL;
	dbtr->db.field = dbALL;
	dbtr->db.record = dbALL;

	dbquery( db, dbDATABASE_NAME, &val );

	strcpy( dbtr->dbname, val.t );

	dbtr->tables = newarr( 0 );

	dbquery( db, dbSCHEMA_TABLES, &schema_tables );

	for( itable = 0; itable < maxtbl( schema_tables ); itable++ ) {

		table_name = gettbl( schema_tables, itable );
	
		ttr = new_tabletrack( table_name );

		setarr( dbtr->tables, ttr->table_name, ttr );

		ttr->db = dblookup( db, "", table_name, "", "" );

		ttr->db.record = dbNULL;

		ttr->null_sync = dbmon_compute_row_sync( ttr->db );

		ttr->db.record = dbALL;
	}

	return dbtr;
}

static Tabletrack *
new_tabletrack( char *table_name )
{
	Tabletrack *ttr = NULL;

	allot( Tabletrack *, ttr, 1 );

	strcpy( ttr->table_name, table_name );

	memset( ttr->table_filename, '\0', FILENAME_MAX );

	ttr->table_exists = 0;
	ttr->table_nrecs = 0L;
	ttr->watch_table = 0;
	ttr->table_modtime = 0L;

	ttr->null_sync = NULL;

	ttr->syncs = newstbl( cmp_synctracks );

	return ttr;
}

static Synctrack *
new_synctrack( Dbptr db, long irecord )
{
	Synctrack *str = NULL;

	allot( Synctrack *, str, 1 );

	str->sync = dbmon_compute_row_sync( db );

	str->irecord = irecord;

	str->keep = 0;
	str->add = 0;

	return str;
}

static SynctrackCtxt *
new_synctrack_context( Dbtrack *dbtr, Tabletrack *ttr, void *pvt ) 
{
	SynctrackCtxt *strcxt = NULL;

	allot( SynctrackCtxt *, strcxt, 1 );

	strcxt->dbtr = dbtr;
	strcxt->ttr = ttr;
	strcxt->pvt = pvt;

	return strcxt;
}

static void
free_synctrack_context( SynctrackCtxt *strp )
{
	free( strp );

	return;
}

static void
free_synctrack( void *strp )
{
	Synctrack *str = (Synctrack *) strp;

	if( str->sync != (char *) NULL ) {

		free( str->sync );
	}

	free( str );

	return;
}

static void
free_tabletrack( void *ttrp )
{
	Tabletrack *ttr = (Tabletrack *) ttrp;

	if( ttr->null_sync != (char *) NULL ) {

		free( ttr->null_sync );
	}

	if( ttr->syncs != (Stbl *) NULL ) {
		
		freestbl( ttr->syncs, free_synctrack );
	}

	free( ttr );

	return;
}

static void
free_dbtrack( void *dbtrp )
{
	Dbtrack *dbtr = (Dbtrack *) dbtrp;

	freearr( dbtr->tables, free_tabletrack );

	free( dbtr );

	return;
}

static int
cmp_synctracks( void *ap, void *bp )
{
	Synctrack *a = (Synctrack *) ap;
	Synctrack *b = (Synctrack *) bp;

	return strcmp( a->sync, b->sync );
}

static int
synctrack_print( void *strp, void *fpp )
{
	Synctrack *str = (Synctrack *) strp;
	FILE	*fp = (FILE *) fpp;

	fprintf( fp, "\t%d\t%s\n", str->irecord, str->sync );

	return 0;
}

static int
synctrack_reset_conditionals( void *strp, void *pvt )
{
	Synctrack *str = (Synctrack *) strp;

	str->add = 0;

	str->keep = 0;

	return 0;
}

static int
synctrack_certain_delete( void *strp, void *strcxtp )
{
	Synctrack *str = (Synctrack *) strp;
	SynctrackCtxt *strcxt = (SynctrackCtxt *) strcxtp;

	strcxt->dbtr->delrow( strcxt->ttr->db, 
		      	      strcxt->ttr->table_name, 
		      	      str->sync, 
		      	      strcxt->pvt );

	delstbl( strcxt->ttr->syncs, str );

	free_synctrack( (void *) str );

	return 0;
}

static int
synctrack_conditional_delete( void *strp, void *strcxtp )
{
	Synctrack *str = (Synctrack *) strp;
	int	rc = 0;

	if( ! str->keep ) {

		rc = synctrack_certain_delete( strp, strcxtp );
	}

	return rc;
}

static int
synctrack_conditional_add( void *strp, void *strcxtp )
{
	Synctrack *str = (Synctrack *) strp;
	SynctrackCtxt *strcxt = (SynctrackCtxt *) strcxtp;
	Dbptr	db;
	Dbptr	dbscratch;
	char	*checksync = NULL;
	int	rc = 0;

	if( ! str->add ) {

		return rc;
	}

	db = strcxt->ttr->db;

	db.record = str->irecord;

	dbget( db, NULL );

	dbscratch = dblookup( db, "", "", "", "dbSCRATCH" );

	checksync = dbmon_compute_row_sync( dbscratch );

	if( strcmp( str->sync, checksync ) ) {

		/* Row changed; leave for next iteration */

		free( checksync );

		return rc;
	}

	free( checksync );

	strcxt->dbtr->newrow( dbscratch, strcxt->ttr->table_name, str->irecord, str->sync, strcxt->pvt );	

	return rc;
}

static int
sort_byirecord( char **ap, char **bp, void *pvt )
{
	Synctrack *a = (Synctrack *) *ap;
	Synctrack *b = (Synctrack *) *bp;
	int	rc;

	if( a->irecord < b->irecord ) {

		rc = -1;

	} else if( a->irecord == b->irecord ) {

		rc = 0;

	} else {

		rc = 1;
	}

	return rc;
}

static int 
applystbl_sorted( Stbl *stbl, int (*afunction)(void *, void *), void *pvt )
{
	Tbl	*tbl;
	int	rc = 0;

	tbl = tblstbl( stbl );

	sorttbl( tbl, sort_byirecord, NULL );

	applytbl( tbl, afunction, pvt );

	return rc;
}

static void
focus_tableset( Dbtrack *dbtr, Tbl *table_subset )
{
	Tbl	*keys = NULL;
	long	ikey = 0L;
	long	itable = 0L;
	char	*table_name = NULL;
	Tabletrack *ttr = NULL;

	if( table_subset == (Tbl *) NULL ) {

		keys = keysarr( dbtr->tables );

		for( ikey = 0; ikey < maxtbl( keys ); ikey++ ) {
		
			ttr = (Tabletrack *) getarr( dbtr->tables, 
						     gettbl( keys, ikey ) );

			ttr->watch_table = 1;
		}

		freetbl( keys, 0 );

	} else {

		for( itable = 0; itable < maxtbl( table_subset ); itable++ ) {

			table_name = gettbl( table_subset, itable );	

			ttr = (Tabletrack *) getarr( dbtr->tables, table_name );

			ttr->watch_table = 1;
		}
	}

	return;
}

static void
dbmon_build_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt )
{
	Dbptr	db;
	Dbptr	dbscratch;
	long	irecord = 0L;
	long	new_nrecs = 0L;
	Synctrack *newstr = NULL;
	SynctrackCtxt *strcxt = NULL;

	db = ttr->db;

	dbscratch = dblookup( db, "", "", "", "dbSCRATCH" );

	dbquery( ttr->db, dbRECORD_COUNT, &new_nrecs );	

	for( irecord = 0; irecord < new_nrecs; irecord++ ) {

		db.record = irecord;

		dbget( db, NULL );

		newstr = new_synctrack( dbscratch, irecord );

		if( ! strcmp( newstr->sync, ttr->null_sync ) ) {

			free_synctrack( newstr );

			continue;
		}

		newstr->keep = 1;

		newstr->add = 1;

		addstbl( ttr->syncs, (void *) newstr );
	}

	strcxt = new_synctrack_context( dbtr, ttr, pvt );

	applystbl_sorted( ttr->syncs, synctrack_conditional_add, (void *) strcxt );

	free_synctrack_context( strcxt );

	return;
}

static void
dbmon_update_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt )
{
	long	irecord = 0L;
	long	new_nrecs = 0L;
	Dbptr	db;
	Dbptr	dbscratch;
	Synctrack *oldstr = NULL;
	Synctrack *newstr = NULL;
	SynctrackCtxt *strcxt = NULL;

	db = ttr->db;

	dbscratch = dblookup( db, "", "", "", "dbSCRATCH" );

	applystbl( ttr->syncs, synctrack_reset_conditionals, NULL );

	/* Use dbquery in attempt to prevent bus error from reading past end 
	   of table that may still be shortening: */

	for( irecord = 0; 
		irecord < dbquery( ttr->db, dbRECORD_COUNT, &new_nrecs ); 
		   irecord++ ) {

		db.record = irecord;

		dbget( db, NULL );

		newstr = new_synctrack( dbscratch, irecord );

		if( ! strcmp( newstr->sync, ttr->null_sync ) ) {

			free_synctrack( newstr );

			continue;
		}

		newstr->keep = 1;

		oldstr = (Synctrack *) tststbl( ttr->syncs, newstr );

		if( oldstr != (Synctrack *) NULL ) {

			delstbl( ttr->syncs, oldstr );

			free_synctrack( (void *) oldstr );

			newstr->add = 0;

		} else {

			newstr->add = 1;
		}

		addstbl( ttr->syncs, (void *) newstr );
	}

	strcxt = new_synctrack_context( dbtr, ttr, pvt );

	applystbl_sorted( ttr->syncs, synctrack_conditional_delete, (void *) strcxt );

	applystbl_sorted( ttr->syncs, synctrack_conditional_add, (void *) strcxt );

	free_synctrack_context( strcxt );
		
	return;
}

static void
dbmon_delete_table( Dbtrack *dbtr, Tabletrack *ttr, void *pvt )
{
	SynctrackCtxt *strcxt = NULL;

	strcxt = new_synctrack_context( dbtr, ttr, pvt );

	applystbl( ttr->syncs, synctrack_certain_delete, (void *) strcxt );

	free_synctrack_context( strcxt );

	return;
}

char *
dbmon_compute_row_sync( Dbptr db )
{
	unsigned long record_size = 0L;
	unsigned char digest[20];
	char	*sync = (char *) NULL;
	char	*row = (char *) NULL;

	db.field = dbALL;

	dbquery( db, dbRECORD_SIZE, &record_size );

	allot( char *, row, record_size + 2 );

	dbget( db, row );

	compute_digest( (unsigned char *) row, record_size, &digest[0] );

	free( row ); 

	sync = digest2hex( digest );

	return sync;
}

Hook *
dbmon_init( Dbptr db, Tbl *table_subset, 
	    void (*newrow)(Dbptr, char *, long, char *, void *), 
	    void (*delrow)(Dbptr, char *, char *, void *), 
	    int flags )
{
	Hook	*dbmon_hook = NULL;
	Dbtrack *dbtr = NULL;

	dbmon_hook = new_hook( free_dbtrack );

	dbtr = new_dbtrack( db );

	focus_tableset( dbtr, table_subset );

	dbtr->newrow = newrow;
	dbtr->delrow = delrow;

	dbmon_hook->p = (void *) dbtr;

	return dbmon_hook;
}

int 
dbmon_update( Hook *dbmon_hook, void *pvt )
{
	Dbtrack *dbtr = (Dbtrack *) dbmon_hook->p;
	Tabletrack *ttr;
	Tbl	*keys;
	Dbvalue val;
	int	retcode = 0;
	long	ikey;
	long	new_nrecs = 0L;
	char	cmd[FILENAME_MAX+STRSZ];

	sprintf( cmd, "orb2db_msg %s wait", dbtr->dbname );

	system( cmd );

	keys = keysarr( dbtr->tables );

	for( ikey = 0; ikey < maxtbl( keys ); ikey++ ) {
		
		ttr = (Tabletrack *) getarr( dbtr->tables, gettbl( keys, ikey ) );

		if( ! ttr->watch_table ) {

			continue;
		}

		dbflush_indexes( ttr->db );

		dbquery( ttr->db, dbRECORD_COUNT, &new_nrecs );	

		if( ttr->table_nrecs == 0 && new_nrecs <= 0 ) { 			/* Table still nonexistent */

			continue;

		} else if( ttr->table_nrecs == 0 && new_nrecs >= 0 ) {			/* Table appeared */

			dbquery( ttr->db, dbTABLE_FILENAME, &val );

			strcpy( ttr->table_filename, val.t );

			dbmon_build_table( dbtr, ttr, pvt );

			ttr->table_exists = 1;

		} else if( ttr->table_nrecs > 0 && new_nrecs <= 0 ) { 			/* Table disappeared */

			dbmon_delete_table( dbtr, ttr, pvt );

			ttr->table_exists = 0;

		} else if( ttr->table_modtime != filetime( ttr->table_filename ) ) { 	/* Table changed */

			dbmon_update_table( dbtr, ttr, pvt );

		} else {								 /* Table unchanged */

			; 	/* Do nothing */
		}
		
		ttr->table_nrecs = new_nrecs;

		ttr->table_modtime = filetime( ttr->table_filename );
	}

	freetbl( keys, 0 );

	return retcode;
}

void 
dbmon_status( FILE *fp, Hook *dbmon_hook ) 
{
	Dbtrack *dbtr = (Dbtrack *) dbmon_hook->p;
	Tabletrack *ttr = NULL;
	Tbl	*keys = NULL;
	long	ikey = 0L;
	char	*s = NULL;

	fprintf( fp, "Monitoring database: '%s'\n", dbtr->dbname );
	fprintf( fp, "Cached database pointer: %ld %ld %ld %ld\n", 
							dbtr->db.database, 
							dbtr->db.table, 
							dbtr->db.field, 
							dbtr->db.record );

	keys = keysarr( dbtr->tables );

	for( ikey = 0; ikey < maxtbl( keys ); ikey++ ) {
		
		ttr = (Tabletrack *) getarr( dbtr->tables, gettbl( keys, ikey ) );

		fprintf( fp, "Schema table '%s':\n", ttr->table_name );

		if( ! ttr->watch_table ) {

			fprintf( fp, "\tWatched: no\n" );

			continue;
		}

		fprintf( fp, "\tFilename: '%s'\n", ttr->table_filename );

		fprintf( fp, "\tExists: %s\n", ttr->table_exists ? "yes" : "no" );

		if( ttr->table_exists ) {
			fprintf( fp, "\tFile modification time: %s\n", s = strtime( (double) ttr->table_modtime ) );
			free( s );
		}

		fprintf( fp, "\tNumber of records: %ld\n", ttr->table_nrecs );

		if( ttr->null_sync != (char *) NULL ) {

			fprintf( fp, "\tNull-row sync string: %s\n", ttr->null_sync );
		}

		if( ttr->syncs != (Stbl *) NULL ) {

			fprintf( fp, "\tSync strings:\n" );

			applystbl_sorted( ttr->syncs, synctrack_print, (void *) fp );
		}

		fprintf( fp, "\tWatched: yes\n" );
	}

	freetbl( keys, 0 );

	return;
}

void 
dbmon_close( Hook **dbmon_hook )
{
	free_hook( dbmon_hook );

	*dbmon_hook = NULL;

	return;
}
